File: Code/C++/Binary_search.cpp
/*
Binary search is an efficient algorithm for finding an item from a sorted list of items. It works by repeatedly dividing in half the portion of the list that could contain the item, until you've narrowed down the possible locations to just one.
We basically ignore half of the elements just after one comparison.

Compare the element to be searched (let be x) with the middle element.
1.If x matches with middle element, we return the mid index.
2.Else If x is greater than the mid element, then x can only lie in right half subarray after the mid element. So we recur for right half.
3.Else (x is smaller) recur for the left half.
*/

#include <bits/stdc++.h>
using namespace std;

int binarySearch(int arr[], int left, int right, int x)
{
	while (left <= right) {
		int mid = left + (right - left) / 2;

		// Check if x is present at mid
		if (arr[mid] == x)
			return mid;

		// If x greater, ignore left half
		if (arr[mid] < x)
			left = mid + 1;

		// If x is smaller, ignore right half
		else
			right = mid - 1;
	}

	// if we reach here, then element was
	// not present
	return -1;
}

int main(void)
{   
	int arr[10];
    cout<<"Enter 10 elements: "<<endl;
    int num = 10;
    for(int i=0;i<num;i++)
    {
        cin>>arr[i];
    }
	int x;
    cout<<"Enter the element to be searched: ";
    cin>>x;
    int n= sizeof(arr)/sizeof(arr[0]);
	int result = binarySearch(arr, 0, n - 1, x);
	(result == -1) ? cout << "Element is not present in array"
				: cout << "Element is present at index " << result;
	return 0;
}
/*
Test Cases:
Input:
Enter 10 elements: 
33
22
11
44
55
66
77
88
55
89
Enter the element to be searched: 66
Output:
Element is present at index 5



Input:
Enter 10 elements: 
34
12
44
54
65
76
87
98
09
21
Enter the element to be searched: 45
output:
Element is not present in array
*/

/*

Time Complexity:O(Log n), which is efficient.
Space Complexity: O(1) in case of iterative implementation.
*/

File: Code/Java/binary_search.java
/*
 * This code is to implement the binary search algorithm.
 * 
 * Binary search is used to search a value in a sorted array
 * by repeatedly dividing the search interval in half and comparing
 * the value of the middle element to the value to be searched.
 * Based on the comparison, the first, last and mid values are updated,
 * by narrowing the interval.
 */

import java.util.*;

public class binary_search
{
    void main()
    {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("Enter number of elements: ");
        int n = scanner.nextInt();
        scanner.nextLine();
        
        int arr[] = new int[n];
        
        System.out.println("Enter elements in sorted order: ");
        
        for(int i = 0; i < n; i++)
        {
            arr[i] = scanner.nextInt();
            scanner.nextLine();
        }
        
        System.out.println("Enter value to be searched: ");
        int value = scanner.nextInt();
        scanner.nextLine();
        
        int first = 0, last = n - 1, mid = 0, flag = 0;
        
        while(first < last)
        {
            mid = (first + last) / 2;
            
            if(value > arr[mid])
            {
                first = mid + 1;
                continue;
            }
            
            else if(value < arr[mid])
            {
                last = mid - 1;
                continue;
            }
            
            else if(value == arr[mid])
            {
                flag = 1;
                break;
            }
        }
        
        if(flag == 1)
        {
            System.out.println("Value found");
        }
        else
        {
            System.out.println("Value not found");
        }
    }
}

/*
 * Test Cases-
 * 
 * 1.
 * Enter number of elements: 
 * 5
 * Enter elements in sorted order: 
 * 1
 * 4
 * 8
 * 10
 * 17
 * Enter value to be searched: 
 * 10
 * Value found
 * 
 * 2.
 * Enter number of elements: 
 * 5
 * Enter elements in sorted order: 
 * 1
 * 6
 * 9
 * 18
 * 55
 * Enter value to be searched: 
 * 1
 * Value found
 * 
 * 3.
 * Enter number of elements: 
 * 5
 * Enter elements in sorted order: 
 * 3
 * 6
 * 9
 * 10
 * 19
 * Enter value to be searched: 
 * 4
 * Value not found
 * 
 * Time Complexity: O(log n)
 * Space Complexity: O(1)
 * where n is the number of elements in the array
 */
